src/live_restore/live_restore_fs.c

```diff
@@ -298,55 +300,76 @@ __live_restore_fs_directory_list_worker(WT_FILE_SYSTEM *fs, WT_SESSION *wt_sessi
-            WT_ERR(__wt_realloc_def(session, &dirallocsz, count_dest + 1, &entries));
-            WT_ERR(__wt_strdup(session, *namep, &entries[count_dest]));
-            ++count_dest;
+    WT_ERR(lr_fs->os_file_system->fs_exist(
+      lr_fs->os_file_system, wt_session, path_dest, &dest_folder_exists));
 
-            if (single)
-                goto done;
-        }
+    if (dest_folder_exists) {
+        WT_ERR(lr_fs->os_file_system->fs_directory_list(
+          lr_fs->os_file_system, wt_session, path_dest, prefix, &dirlist_dest, &num_dest_files));
+
+        for (namep = dirlist_dest; namep != NULL && *namep != NULL; namep++)
+            if (!WT_SUFFIX_MATCH(*namep, WT_LIVE_RESTORE_FS_TOMBSTONE_SUFFIX)) {
+                WT_ERR(__wt_realloc_def(session, &dirallocsz, count_dest + 1, &entries));
+                WT_ERR(__wt_strdup(session, *namep, &entries[count_dest]));
+                ++count_dest;
+
+                if (single)
+                    goto done;
+            }
+    }
 
     /* Get files from source. */
     WT_ERR(__live_restore_fs_backing_filename(
       &lr_fs->source, session, lr_fs->destination.home, directory, &path_src));
-    WT_ERR_ERROR_OK(lr_fs->os_file_system->fs_directory_list(
-                      lr_fs->os_file_system, wt_session, path_src, prefix, &dirlist_src, countp),
-      ENOENT, false);
 
-    for (namep = dirlist_src; namep != NULL && *namep != NULL; namep++) {
-        /* Create file path for the file from the source directory. */
-        WT_ERR(__live_restore_create_file_path(session, &lr_fs->source, *namep, &temp_path));
-        WT_ERR_NOTFOUND_OK(
-          __live_restore_fs_has_file(lr_fs, &lr_fs->destination, session, temp_path, &dest_exist),
-          false);
-        WT_ERR(__dest_has_tombstone(lr_fs, temp_path, session, &have_tombstone));
-
-        if (!dest_exist && !have_tombstone) {
-            WT_ERR(__wt_realloc_def(session, &dirallocsz, count_dest + count_src + 1, &entries));
-            WT_ERR(__wt_strdup(session, *namep, &entries[count_dest + count_src]));
-            ++count_src;
-        }
+    WT_ERR(lr_fs->os_file_system->fs_exist(
+      lr_fs->os_file_system, wt_session, path_src, &source_folder_exists));
 
-        __wt_free(session, temp_path);
-        if (single)
-            goto done;
+    if (source_folder_exists) {
+        WT_ERR(lr_fs->os_file_system->fs_directory_list(
+          lr_fs->os_file_system, wt_session, path_src, prefix, &dirlist_src, &num_src_files));
+
+        for (namep = dirlist_src; namep != NULL && *namep != NULL; namep++) {
+            /*
+             * We're iterating files in the source, but we want to check if they exist in the
+             * destination, so create the file path to the backing destination file.
+             */
+            WT_ERR(
+              __live_restore_create_file_path(session, &lr_fs->destination, *namep, &temp_path));
+            WT_ERR_NOTFOUND_OK(__live_restore_fs_has_file(
```

Comment:
Could we skip these two checks if `dest_folder_exists` is false?

Reply:
We can, but I don't think we get much benefit from it. It'd save us from a few string allocation but directory_list is on the cold path, and trying out the change locally it makes the code a bit harder to read with the extra branching/indentation.

Reply:
I know most programmers are indentationphobes but I think the extra if statement is probably worth it.

Reply:
[0b2094c](https://github.com/wiredtiger/wiredtiger/pull/11369/commits/0b2094cd02acfd24efa9be4bb7546526ad980e02)